### PR TITLE
Built-in per-route monitoring thanks to VaporMonitoring and SwiftPrometheus

### PR DESCRIPTION
**Pull request by Yasumoto**
This will look like (on `/metrics`):

```
# TYPE http_requests_total counter
http_requests_total 0
http_requests_total{status_code="500", path="/status", method="GET"} 2
http_requests_total{status_code="200", path="/metrics", method="GET"} 3
http_requests_total{status_code="500", path="/", method="GET"} 4
http_requests_total{status_code="200", path="/health", method="GET"} 2
# TYPE http_requests_duration_seconds summary
http_requests_duration_seconds{quantile="0.01"} 0.000259041
http_requests_duration_seconds{quantile="0.05"} 0.000259041
http_requests_duration_seconds{quantile="0.5"} 0.289999008
http_requests_duration_seconds{quantile="0.9"} 0.310270905
http_requests_duration_seconds{quantile="0.95"} 0.340256094
http_requests_duration_seconds{quantile="0.99"} 0.340256094
http_requests_duration_seconds{quantile="0.999"} 0.340256094
http_requests_duration_seconds_count 11
http_requests_duration_seconds_sum 1
http_requests_duration_seconds{status_code="200", quantile="0.01", path="/metrics", method="GET"} 0.000995993
http_requests_duration_seconds{status_code="200", quantile="0.05", path="/metrics", method="GET"} 0.000995993
http_requests_duration_seconds{status_code="200", quantile="0.5", path="/metrics", method="GET"} 0.00285995
http_requests_duration_seconds{status_code="200", quantile="0.9", path="/metrics", method="GET"} 0.002887964
http_requests_duration_seconds{status_code="200", quantile="0.95", path="/metrics", method="GET"} 0.002887964
http_requests_duration_seconds{status_code="200", quantile="0.99", path="/metrics", method="GET"} 0.002887964
http_requests_duration_seconds{status_code="200", quantile="0.999", path="/metrics", method="GET"} 0.002887964
http_requests_duration_seconds_count{status_code="200", path="/metrics", method="GET"} 3
http_requests_duration_seconds_sum{status_code="200", path="/metrics", method="GET"} 0
http_requests_duration_seconds{status_code="500", quantile="0.01", path="/", method="GET"} 0.293069005
http_requests_duration_seconds{status_code="500", quantile="0.05", path="/", method="GET"} 0.293069005
http_requests_duration_seconds{status_code="500", quantile="0.5", path="/", method="GET"} 0.3019359705
http_requests_duration_seconds{status_code="500", quantile="0.9", path="/", method="GET"} 0.340256094
http_requests_duration_seconds{status_code="500", quantile="0.95", path="/", method="GET"} 0.340256094
http_requests_duration_seconds{status_code="500", quantile="0.99", path="/", method="GET"} 0.340256094
http_requests_duration_seconds{status_code="500", quantile="0.999", path="/", method="GET"} 0.340256094
http_requests_duration_seconds_count{status_code="500", path="/", method="GET"} 4
http_requests_duration_seconds_sum{status_code="500", path="/", method="GET"} 1
http_requests_duration_seconds{status_code="500", quantile="0.01", path="/status", method="GET"} 0.289999008
http_requests_duration_seconds{status_code="500", quantile="0.05", path="/status", method="GET"} 0.289999008
http_requests_duration_seconds{status_code="500", quantile="0.5", path="/status", method="GET"} 0.289999008
http_requests_duration_seconds{status_code="500", quantile="0.9", path="/status", method="GET"} 0.291983962
http_requests_duration_seconds{status_code="500", quantile="0.95", path="/status", method="GET"} 0.291983962
http_requests_duration_seconds{status_code="500", quantile="0.99", path="/status", method="GET"} 0.291983962
http_requests_duration_seconds{status_code="500", quantile="0.999", path="/status", method="GET"} 0.291983962
http_requests_duration_seconds_count{status_code="500", path="/status", method="GET"} 2
http_requests_duration_seconds_sum{status_code="500", path="/status", method="GET"} 0
http_requests_duration_seconds{status_code="200", quantile="0.01", path="/health", method="GET"} 0.000259041
http_requests_duration_seconds{status_code="200", quantile="0.05", path="/health", method="GET"} 0.000259041
http_requests_duration_seconds{status_code="200", quantile="0.5", path="/health", method="GET"} 0.000259041
http_requests_duration_seconds{status_code="200", quantile="0.9", path="/health", method="GET"} 0.000380992
http_requests_duration_seconds{status_code="200", quantile="0.95", path="/health", method="GET"} 0.000380992
http_requests_duration_seconds{status_code="200", quantile="0.99", path="/health", method="GET"} 0.000380992
http_requests_duration_seconds{status_code="200", quantile="0.999", path="/health", method="GET"} 0.000380992
http_requests_duration_seconds_count{status_code="200", path="/health", method="GET"} 2
http_requests_duration_seconds_sum{status_code="200", path="/health", method="GET"} 0
```

### CODE REVIEW

File: Sources/App/configure.swift [hunk 1]
```diff
@@ -32,7 +37,10 @@ public func configure(_ config: inout Config, _ env: inout Environment, _ servic
     services.register(databases)
 
     // Register middleware
+    services.register(MetricsMiddleware(), as: MetricsMiddleware.self)
+
     var middlewares = MiddlewareConfig() // Create _empty_ middleware config
+    middlewares.use(MetricsMiddleware.self) // We want to see RED (Requests, Errors, Duration) for all requests
```
Comment (fxchen) [inline Sources/App/configure.swift:43]: I like this explicitness.

File: Sources/App/configure.swift [hunk 2]
```diff
@@ -12,6 +13,10 @@ public func configure(_ config: inout Config, _ env: inout Environment, _ servic
     let router = EngineRouter.default()
     try routes(router)
     services.register(router, as: Router.self)
+
+    // Use Prometheus 🔥 for monitoring
+    let prometheusService = VaporPrometheus(router: router, route: "metrics")
```
Comment (fxchen) [inline Sources/App/configure.swift:18]: I just did some reading on https://github.com/vapor-community/VaporMonitoring

💯

File: Package.resolved [hunk 3]
```diff
@@ -253,6 +253,15 @@
           "version": "3.3.0"
         }
       },
+      {
+        "package": "VaporMonitoring",
+        "repositoryURL": "https://github.com/Yasumoto/VaporMonitoring.git",
```
Comment (fxchen) [inline Package.resolved:258]: [Non-blocking comment] How do you plan to perform syncs / updates of this repo?

**Review by fxchen — APPROVED** [review]
Dig this update to include nice request metrics.

File: Package.resolved (hunk 3 above)
Comment (Yasumoto) [reply Package.resolved]: Once we get the upstream PR merged (at https://github.com/vapor-community/VaporMonitoring/pull/5) then we'll revert back to upstream.

Comment (Yasumoto) [reply Package.resolved:258]: Once we get the upstream PR merged (at vapor-community/VaporMonitoring#5) then we'll revert back to upstream.

**Pull request merged**